{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script>requirejs.config({paths: { 'plotly': ['https://cdn.plot.ly/plotly-latest.min']},});if(!window.Plotly) {{require(['plotly'],function(plotly) {window.Plotly=plotly;});}}</script>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['c', 't', 'g', 'a', 'a', 'c', 'c', 'c', 't', 'a', 'a', 'a', 'n', 'n', 'c', 'c', 'c', 't', 'g', 'a', 'a', 'g', 'c', 'c', 'c', 'c', 't', 't', 'c', 'a']\n",
      "43.333333333333336\n",
      "[43.333333333333336]\n",
      "[ 43.33333333]\n",
      "[]\n",
      "[ 43.33333333]\n",
      "[]\n",
      "[ 43.33333333]\n",
      "[]\n",
      "[ 43.33333333]\n",
      "[]\n",
      "[ 43.33333333]\n",
      "[]\n",
      "[ 43.33333333]\n",
      "[]\n",
      "[ 43.33333333]\n",
      "[]\n",
      "[ 43.33333333]\n",
      "[]\n",
      "[ 43.33333333]\n",
      "[]\n",
      "[ 43.33333333]\n",
      "[]\n",
      "[ 43.33333333]\n",
      "[]\n",
      "[ 43.33333333]\n",
      "['c', 'c', 'c', 't', 'g', 'a', 'a', 'c', 'c', 'c', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 't', 'a', 'a', 'a', 'c', 'c']\n",
      "23.333333333333332\n",
      "['c', 'c', 't', 'g', 'a', 'a', 'c', 'c', 'c', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 't', 'a', 'a', 'a', 'c', 'c', 'c']\n",
      "23.333333333333332\n",
      "['c', 't', 'g', 'a', 'a', 'c', 'c', 'c', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 't', 'a', 'a', 'a', 'c', 'c', 'c', 't']\n",
      "26.666666666666668\n",
      "['t', 'g', 'a', 'a', 'c', 'c', 'c', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 't', 'a', 'a', 'a', 'c', 'c', 'c', 't', 'a']\n",
      "30.0\n",
      "['g', 'a', 'a', 'c', 'c', 'c', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 't', 'a', 'a', 'a', 'c', 'c', 'c', 't', 'a', 'n']\n",
      "26.666666666666668\n",
      "[23.333333333333332, 23.333333333333332, 26.666666666666668, 30.0, 26.666666666666668]\n",
      "[ 23.33333333  23.33333333  26.66666667  30.          26.66666667]\n",
      "[]\n",
      "[ 23.33333333  23.33333333  26.66666667  30.          26.66666667]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([ 23.33333333,  23.33333333,  26.66666667,  30.        ,  26.66666667])"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#SynMap2GCcontent\n",
    "##Step 1 Importing FASTA files\n",
    "\n",
    "####Import library\n",
    "import os\n",
    "from itertools import islice\n",
    "from itertools import chain, repeat\n",
    "import numpy as np\n",
    "from plotly import __version__\n",
    "from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot\n",
    "from plotly.graph_objs import Bar, Scatter, Figure, Layout\n",
    "init_notebook_mode(connected=True)\n",
    "\n",
    "\n",
    "####Global variables \n",
    "header = []\n",
    "nucleotides = []\n",
    "window_size = 30\n",
    "\n",
    "\n",
    "####File input/path\n",
    "#my_file = open(os.path.expanduser('Plasmodium_chabaudi_chabaudi_strain_AS.faa'))\n",
    "my_file = open(os.path.expanduser('Plasmodium__test.faa'))\n",
    "#my_file = open(os.path.expanduser('minitestdata.fasta'))\n",
    "#my_file = open(os.path.expanduser('Plasmodium_chabaudi_chabaudi_TEST2.faa'))\n",
    "file_contents = my_file.read()\n",
    "file_contents_lower = file_contents.lower()                                                    #Lowercases all strings in file\n",
    "file_contents_lower_split = file_contents_lower.split()                                        #Split the whole file string into words\n",
    "                                                                       \n",
    "\n",
    "##Step 2 Separating the file from a single string into two lists    \n",
    "for i in file_contents_lower_split:                                                            #Divides words into two lists: one with the simbol \">\" and the other one without it\n",
    "    if \">\" in i:\n",
    "        header.append(i)\n",
    "    else:\n",
    "        nucleotides.append(i)\n",
    "\n",
    "\n",
    "def window(seq, window_size):\n",
    "#Makes a sliding window in the sequence\n",
    "    foriter = []\n",
    "    for n in seq:\n",
    "        foriter.append(n)\n",
    "    #print(foriter)\n",
    "    lst = iter(foriter)\n",
    "    result = tuple(islice(lst, window_size))\n",
    "    if len(result) == window_size:\n",
    "        yield list(result)    \n",
    "    for elem in lst:\n",
    "        result = result[1:] + (elem,)\n",
    "        yield list(result)\n",
    "\n",
    "\n",
    "# def GC_counter(seq):\n",
    "# #Performs a rolling calculation of GC% for the sequence in the corresponding sequence segment\n",
    "#     seq_tuple = window(seq, window_size)\n",
    "#     frag_list_gc = []\n",
    "#     frag_lenghts = []\n",
    "#     for n in seq_tuple:    \n",
    "#         l = len(n)\n",
    "#         frag_lenghts.append(l)\n",
    "#         count_g = n.count('g')\n",
    "#         count_c = n.count('c')\n",
    "#         count_gc = count_g+count_c\n",
    "#         percent_gc = (float(count_gc/l))*100\n",
    "#         frag_list_gc.append(percent_gc)\n",
    "#         percent_array_gc = np.asarray(frag_list_gc)\n",
    "#         global percent_array_gc\n",
    "#     return \n",
    "#     #return percent_array_gc\n",
    "\n",
    "\n",
    "# def N_counter(seq):\n",
    "# #Performs a rolling calculation of N% for the sequence in the corresponding sequence segment\n",
    "#     seq_tuple = window(seq, window_size)\n",
    "#     frag_list_n = []\n",
    "#     frag_lenghts = []\n",
    "#     for n in seq_tuple:    \n",
    "#         l = len(n)\n",
    "#         frag_lenghts.append(l)\n",
    "#         count_n = n.count('n')\n",
    "#         percent_n = (float(count_n/l))*100\n",
    "#         frag_list_n.append(percent_n)\n",
    "#         percent_array_n = np.asarray(frag_list_n)\n",
    "#         #global percent_array_n\n",
    "#     #return   \n",
    "#     return percent_array_n\n",
    "\n",
    "\n",
    "# def X_counter(seq):\n",
    "# #Performs a rolling calculation of X% for the sequence in the corresponding sequence segment\n",
    "#     seq_tuple = window(seq, window_size)\n",
    "#     frag_list_x = []\n",
    "#     frag_lenghts = []\n",
    "#     for n in seq_tuple:    \n",
    "#         l = len(n)\n",
    "#         frag_lenghts.append(l)\n",
    "#         count_x = n.count('x')\n",
    "#         percent_x = (float(count_x/l))*100\n",
    "#         frag_list_x.append(percent_x)\n",
    "#         percent_array_x = np.asarray(frag_list_x)\n",
    "#         #global percent_array_x\n",
    "#     #return \n",
    "#     return percent_array_x\n",
    "\n",
    "\n",
    "# def positioner(seq):\n",
    "# #Loops for every nucleotide per sequence, counts the number of positions and adds them to the position list\n",
    "#     l = 0\n",
    "#     position = []\n",
    "#     for l in range(0,len(seq)):                                                              \n",
    "#         l = l+1\n",
    "#         position.append(l)\n",
    "#         position_array = np.asarray(position)\n",
    "#     return position_array\n",
    "\n",
    "\n",
    "# def dictionary(seq):\n",
    "#         AT_cnt_arr = AT_counter(seq)\n",
    "#         GC_cnt_arr = GC_counter(seq)\n",
    "#         N_cnt_arr = N_counter(seq)\n",
    "#         X_cnt_arr = X_counter(seq)\n",
    "#     for seq in nucleotides:        \n",
    "#         dictio_fractions = dict(zip(header, AT_cnt_arr, GC_cnt_arr, N_cnt_arr, X_cnt_arr))                                             #Combines sequence/chromosome headers and the calculated nucleotide fractions into a dictionary\n",
    "#         print(dictio_fractions)\n",
    "#         return dictio_fractions\n",
    "\n",
    "# dictionary(nucleotides)\n",
    "\n",
    "\n",
    "# def plotly(nucleo):\n",
    "#     for seq in nucleotides:\n",
    "#         AT_cnt_arr = AT_counter(seq)\n",
    "#         GC_cnt_arr = GC_counter(seq)\n",
    "#         N_cnt_arr = N_counter(seq)\n",
    "#         X_cnt_arr = X_counter(seq)\n",
    "#         pos_arr = positioner(seq)\n",
    "    \n",
    "#         AT_trace = Scatter(\n",
    "#         y = AT_cnt_arr,\n",
    "#         x = pos_arr,\n",
    "#         name='AT content',\n",
    "#         line = dict(color = ('rgb(3,141,243)'))) #Blue\n",
    "        \n",
    "#         GC_trace = Scatter(\n",
    "#         y = GC_cnt_arr,\n",
    "#         x = pos_arr,\n",
    "#         name='GC content',\n",
    "#         line = dict(color = ('rgb(64,182,77)'))) #Green\n",
    "        \n",
    "#         N_trace = Scatter(\n",
    "#         y = N_cnt_arr,\n",
    "#         x = pos_arr,\n",
    "#         name='N content',\n",
    "#         line = dict(color = ('rgb(243,145,3)'))) #Orange\n",
    "        \n",
    "#         X_trace = Scatter(\n",
    "#         y = X_cnt_arr,\n",
    "#         x = pos_arr,\n",
    "#         name='X content',\n",
    "#         line = dict(color = ('rgb(171,3,243)'))) #Purple\n",
    "        \n",
    "        \n",
    "#         data = [AT_trace, GC_trace, N_trace, X_trace]  \n",
    "\n",
    "#         layout = dict(xaxis=dict(title='Window iteration',rangeslider=dict()),yaxis=dict(title='Percentage'))\n",
    "#         fig = dict(data=data, layout=layout)\n",
    "#         iplot(fig)              \n",
    "#     return\n",
    "\n",
    "\n",
    "#plotly(nucleotides)\n",
    "\n",
    "\n",
    "        \n",
    "def AT_counter(seq):\n",
    "#Performs a rolling calculation of AT% for the sequence in the corresponding sequence segment\n",
    "    for seq in nucleotides:\n",
    "        #print (seq)\n",
    "        seq_tuple = window(seq, window_size)\n",
    "        frag_list_at = []\n",
    "        frag_lenghts = []\n",
    "        for n in seq_tuple:\n",
    "            print(n)\n",
    "            l = len(n)\n",
    "            frag_lenghts.append(l)\n",
    "            count_a = n.count('t')\n",
    "            count_t = n.count('a')\n",
    "            count_at = count_a+count_t\n",
    "            percent_at = (float(count_at/l))*100\n",
    "            print (percent_at)\n",
    "            frag_list_at.append(percent_at)\n",
    "            percent_array_at = np.asarray(frag_list_at)\n",
    "            #global percent_array_at\n",
    "        #return \n",
    "        print (frag_list_at)\n",
    "        print (percent_array_at)\n",
    "    return percent_array_at\n",
    "\n",
    "AT_counter(nucleotides)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'>lcl|:_2': [57.89, 0.0, 42.11, 0.0, 0.0, 0.0], '>lcl|:_1': [26.67, 40.0, 16.67, 10.0, 6.67, 0.0]}\n"
     ]
    }
   ],
   "source": [
    "#SynMap2GCcontent\n",
    "##Step 1 Importing FASTA files\n",
    "\n",
    "####Import library\n",
    "import os\n",
    "\n",
    "####Global variables \n",
    "header = []\n",
    "nucleotides = []\n",
    "all_fractions = []\n",
    "\n",
    "####Paths\n",
    "#my_file = open(os.path.expanduser('Plasmodium_chabaudi_chabaudi_strain_AS.faa'))\n",
    "#my_file = open(os.path.expanduser('Plasmodium_ovale_curtisi.faa'))\n",
    "my_file = open(os.path.expanduser('minitestdata.fasta'))\n",
    "\n",
    "file_contents = my_file.read()\n",
    "file_contents_lower = file_contents.lower()                                                    #Lowercases all strings in file\n",
    "file_contents_lower_split = file_contents_lower.split()                                        #Split the whole file string into words\n",
    "                                                                       \n",
    "\n",
    "for i in file_contents_lower_split:                                                            #Divides words into two lists: one with the simbol \">\" and the other one without it\n",
    "    if \">\" in i:\n",
    "        header.append(i)\n",
    "    else:\n",
    "        nucleotides.append(i)\n",
    "\n",
    "dictio = dict(zip(header, nucleotides))                                                        #Makes a dictionary using the two created lists: header and nucleotides\n",
    "\n",
    "\n",
    "for seq in nucleotides:                                                                        #Loops for every sequence in the nucleotide list\n",
    "    seq_frac = []\n",
    "    count_a = seq.count('a')                                                                   #Counts the number of each nucleotide per sequence/chromosome in nucleoide list\n",
    "    count_t = seq.count('t')\n",
    "    count_c = seq.count('c')\n",
    "    count_g = seq.count('g')\n",
    "    count_n = seq.count('n')\n",
    "    count_x = seq.count('x')\n",
    "\n",
    "    frac_a = (float(count_a/len(seq)))*100                                                     #Divides the counted nucleotide by the sequence lenght and multiplies it by 100 (created percentage)\n",
    "    frac_c = (float(count_c/len(seq)))*100\n",
    "    frac_t = (float(count_t/len(seq)))*100\n",
    "    frac_g = (float(count_g/len(seq)))*100\n",
    "    frac_n = (float(count_n/len(seq)))*100\n",
    "    frac_x = (float(count_x/len(seq)))*100\n",
    "\n",
    "    seq_frac.append(round(frac_a,2))                                                           #Appends each count into a list unique per sequence/chromosome in nucleotides list. It rounds the decimal point to two\n",
    "    seq_frac.append(round(frac_c,2))\n",
    "    seq_frac.append(round(frac_t,2))\n",
    "    seq_frac.append(round(frac_g,2))\n",
    "    seq_frac.append(round(frac_n,2))\n",
    "    seq_frac.append(round(frac_x,2))\n",
    "\n",
    "    all_fractions.append(seq_frac)                                                              #Appends all nucleotide fractions/percentages to a single list\n",
    "\n",
    "dictio_fractions = dict(zip(header, all_fractions))                                             #Combines sequence/chromosome headers and the calculated nucleotide fractions into a dictionary\n",
    "print(dictio_fractions)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [conda env:py35]",
   "language": "python",
   "name": "conda-env-py35-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
